character-sfx.js

```diff
@@ -308,6 +346,67 @@ class CharacterSfx {
     };
     _handleFood();
   }
+  playGrunt(type, index){
+
+    let voiceFiles, offset, duration;
+    switch (type) {
+      case 'pain': {
+        voiceFiles = this.player.voicePack.actionVoices.filter(f => /Pain/.test(f.name));
+        break;
+      }
+      case 'scream': {
+        voiceFiles = this.player.voicePack.actionVoices.filter(f => /Scream/.test(f.name));
+        break;
+      }
+      case 'attack': {
+        voiceFiles = this.player.voicePack.actionVoices.filter(f => /Attack/.test(f.name));
+        break;
+      }
+      case 'angry': {
+        voiceFiles = this.player.voicePack.actionVoices.filter(f => /Angry/.test(f.name));
+        break;
+      }
+      case 'gasp': {
+        voiceFiles = this.player.voicePack.actionVoices.filter(f => /gasp/.test(f.name));
+        break;
+      }
+      case 'jump': {
+        voiceFiles = this.player.voicePack.actionVoices.filter(f => /jump/.test(f.name));
+        break;
+      }
+      case 'narutoRun': {
+        voiceFiles = this.player.voicePack.actionVoices.filter(f => /nr/.test(f.name));
+        break;
+      }
+    }
+    
+    if(index===undefined){
+      let voice = this.player.voicePack.selectVoice(voiceFiles);
+      duration = voice.duration;
+      offset = voice.offset;
+    }
+    else{
+      duration = voiceFiles[index].duration;
+      offset = voiceFiles[index].offset;
+    } 
+    
+    const audioContext = Avatar.getAudioContext();
+    const audioBufferSourceNode = audioContext.createBufferSource();
+    audioBufferSourceNode.buffer = this.player.voicePack.audioBuffer;
+
+    // control mouth movement with audio volume
+    if (!this.player.avatar.isAudioEnabled()) {
+      this.player.avatar.setAudioEnabled(true);
+    }
+    audioBufferSourceNode.connect(this.player.avatar.getAudioInput());
+
+    if(this.oldGrunt){
+      this.oldGrunt.stop();
+    }
+    this.oldGrunt=audioBufferSourceNode;

```

Comment:
For cleanliness purposes I would wait for the `end` of the buffer and then set to `null` to clear it. Otherwise you are infinitely storing the last grunt long after it is done, which is a memory leak.

Reply:
Same as here: https://github.com/webaverse/app/pull/2835/files#diff-62b89e43ead1a304ba4d7c9b9053d6da5a55029f73da7664fa3a73e30d2993e2R249

Reply:
In this situation, If the oldGrunt is playing, should I stop the `oldGrunt` and play the new grunt like this:

https://user-images.githubusercontent.com/60634884/164500226-d1a36774-c6e9-463c-adbc-ffb75f45126f.mp4

Or should I ban the new grunt if the `oldGrunt` hasn't finished like this:

https://user-images.githubusercontent.com/60634884/164500394-02099364-d063-4997-8758-e9567944608e.mp4

Reply:
I think the first one.

Reply:
change to first one: https://github.com/webaverse/app/pull/2835/commits/87aef8e77eb6cc546d70ac342de266f40490c990